Adding a blog post for windows memory pressure eviction

marosset opened the pull request:
<!--
 Hello!

 PLEASE title the FIRST commit appropriately, so that if you squash all
 your commits into one, the combined commit message makes sense.
 For overall help on editing and submitting pull requests, visit:
  https://kubernetes.io/docs/contribute/suggesting-improvements/

 Use the default base branch, “main”, if you're documenting existing
 features in the English localization.

 If you're working on a different localization (not English), see
 https://kubernetes.io/docs/contribute/new-content/overview/#choose-which-git-branch-to-use
 for advice.

 If you're documenting a feature that will be part of a future release, see
 https://kubernetes.io/docs/contribute/new-content/new-features/ for advice.
-->
### Description

<!--
 Remember to ADD A DESCRIPTION and delete this note before submitting
 your pull request. The description should explain what will change,
 and why.
-->

Support for memory-pressure based eviction for Windows nodes was added in K8s v1.31. 
Documentation on k8s.io website was added with https://github.com/kubernetes/website/pull/47306 but would also like to publish a blog post to cover some of the more Windows specific behaviors

### Issue

<!--
 If this pull request resolves an open issue, please link the issue in the PR
 description so it will automatically close when the PR is merged.

 See the GitHub documentation for more details and other options:

 https://docs.github.com/en/issues/tracking-your-work-with-issues/linking-a-pull-request-to-an-issue#linking-a-pull-request-to-an-issue-using-a-keyword
-->

/sig windows

marosset commented:
/assign @jsturtevant for technical review

sftim commented on content/en/blog/_posts/Windows-Memory-Pressure-Eviction.md:
```diff
@@ -0,0 +1,85 @@
+layout: blog
+title: "Windows Memory Pressure Eviction"
+slug: windows-memory-pressure-eviction
+canonicalUrl: TBD
```
```suggestion
```
No need for this.

sftim reviewed (commented):
Hi. To make life easier for blog reviewers, can you look at https://kubernetes.io/docs/contribute/style/style-guide/ and adopt any guidance that is obviously also relevant for blog articles?

For example, headings should be sentence case, and new concepts should be italicized.

bridgetkromhout commented on content/en/blog/_posts/Windows-Memory-Pressure-Eviction.md:
```diff
@@ -0,0 +1,85 @@
+
+In Kubernetes 1.31, support for memory-pressure based eviction was added for Windows nodes.
+
+This post describes what these changes mean for you and how they may differ from usage on Linux nodes.
```
I parse this as "This post describes what these changes mean for you and how **these changes** may differ from usage on Linux nodes.". If "they" means something else like "Windows node usage", please be explicit.

bridgetkromhout commented on content/en/blog/_posts/Windows-Memory-Pressure-Eviction.md:
```diff
@@ -0,0 +1,85 @@
+
+## Additional Resources
+
+To gain a deeper understanding of how memory is managed on Windows, you may also find this article on [Physical and Virtual Memory in Windows 10](https://answers.microsoft.com/en-us/windows/forum/all/physical-and-virtual-memory-in-windows-10/e36fb5bc-9ac8-49af-951c-e7d39b979938) helpful.
```
Given that Windows 11 is available, does this Windows 10 article from 2015 apply to the newer version also? Or is it scoped only to 10, and maybe 11 has some other info that we need to point to?

marosset replied:
It is still relevant and is the most informative article I could find on the topic

jsturtevant replied:
That is one detailed blog post 😮

jsturtevant commented on content/en/blog/_posts/Windows-Memory-Pressure-Eviction.md:
```diff
@@ -0,0 +1,85 @@
+
+## Eviction Thresholds
+
+By default, Windows nodes have a default "hard" eviction threadhold of `--eviction-hard=memory.available<500Mi`. This means that if the system's availalbe memory is less than 500 Mi the Kubelet will try and start evicting pods.
```
```suggestion
By default, Windows nodes have a default "hard" eviction threadhold of 500Mi (kubelet setting `--eviction-hard=memory.available<500Mi>`). This means that if the system's available commit memory is less than 500 Mi the Kubelet will try and start evicting pods.
```

jsturtevant replied:
this still looks off, specifically the spelling and the missing bracket

marosset replied:
the bracket isn't missing, that is supposed to be a less-than operator.
I have fixed the spelling mistakes and added some clarity tho

jsturtevant commented on content/en/blog/_posts/Windows-Memory-Pressure-Eviction.md:
```diff
@@ -0,0 +1,85 @@
+
+Additionally, exposing the Commit usage for individial Pods and Containers through the `/stats/summary` API is being worked on as part of the [cAdvisor-les, CRI-full Container and Pod Stats](https://github.com/kubernetes/enhancements/blob/master/keps/sig-node/2371-cri-pod-container-stats/README.md) enhancement.
+
+[windows-exporter](https://github.com/prometheus-community/windows_exporter) can also be configured to monitor and report on Commit memory usage.
```
maybe mention it works for containers and node level today if that information is required now?

marosset replied:
updated - PTAL

jsturtevant reviewed (commented):
Overall content looks good from sig-windows perspective

ArvindParekh commented on content/en/blog/_posts/Windows-Memory-Pressure-Eviction.md:
```diff
@@ -0,0 +1,85 @@
+
+Kubernetes nodes can be configured to report memory pressure when the available memory falls below a defined threshold (either a percentage of the total node memory or a specific value). When this signal is raised, the kubelet attempts to evict pods to maintain stability on the node.
+
+For Windows, the available memory is derived from the node's global memory commit levels, which are queried through the [GetPerformanceInfo()](https://](https://learn.microsoft.com/windows/win32/api/psapi/nf-psapi-getperformanceinfo)) system call. Available memory is calculated by subtracting the node's global `CommitTotal` (the amount of memory committed by all processes on the node) from the `C ...
```
```suggestion
For Windows, the available memory is derived from the node's global memory commit levels, which are queried through the [GetPerformanceInfo()](https://learn.microsoft.com/windows/win32/api/psapi/nf-psapi-getperformanceinfo) system call. Available memory is calculated by subtracting the node's global `CommitTotal` (the amount of memory committed by all processes on the node) from the `CommitLimit` (the total amount of virtual memory that can be committed without extending the page file(s) on the node).
```

ArvindParekh commented on content/en/blog/_posts/Windows-Memory-Pressure-Eviction.md:
```diff
@@ -0,0 +1,85 @@
+
+Before diving into the reasons for using commit memory on Windows for eviction signalling, it's helpful to understand the difference between **Working Set** and **Commit** memory.
+
+**Working Set** memory represents the amount of physical memory currently used by processes. It includes the pages of memory that are actively in use and exludes memory pages that have been swapped out to disk (if paging is enabled).
```
```suggestion
**Working Set** memory represents the amount of physical memory currently used by processes. It includes the pages of memory that are actively in use and excludes memory pages that have been swapped out to disk (if paging is enabled).
```

ArvindParekh commented on content/en/blog/_posts/Windows-Memory-Pressure-Eviction.md:
```diff
@@ -0,0 +1,85 @@
+Using global commit levels to measure available memory on Windows was implemented for two key reasons:
+
+1. **Memory limits for containers in Pods running on Windows nodes are enforced using Commit** (not WorkingSet) **memory**. Windows memory management enforces limits on processes based on the total amount of memory committed, not just the memory in use. 
+This also means that memory limits specified for containers in Kubernetes Pods are enforced on the total amount of virtual memory committed by all processes in the container. Consistenty in how memory usage is tracked and enforced is essential for workload stability.
```
```suggestion
This also means that memory limits specified for containers in Kubernetes Pods are enforced on the total amount of virtual memory committed by all processes in the container. Consistency in how memory usage is tracked and enforced is essential for workload stability.
```

ArvindParekh commented on content/en/blog/_posts/Windows-Memory-Pressure-Eviction.md:
```diff
@@ -0,0 +1,85 @@
+
+## Eviction Thresholds
+
+By default, Windows nodes have a default "hard" eviction threadhold of `--eviction-hard=memory.available<500Mi`. This means that if the system's availalbe memory is less than 500 Mi the Kubelet will try and start evicting pods.
```
```suggestion
By default, Windows nodes have a default "hard" eviction threshold of `--eviction-hard=memory.available<500Mi`. This means that if the system's availalbe memory is less than 500 Mi the Kubelet will try and start evicting pods.
```

ArvindParekh commented on content/en/blog/_posts/Windows-Memory-Pressure-Eviction.md:
```diff
@@ -0,0 +1,85 @@
+
+{
+ "node": {
+  "nodeName": "{node-name},
```
```suggestion
  "nodeName": "{node-name}",
```

ArvindParekh reviewed (commented):
Just a few formatting and grammatical errors :)

marosset commented:
> Hi. To make life easier for blog reviewers, can you look at https://kubernetes.io/docs/contribute/style/style-guide/ and adopt any guidance that is obviously also relevant for blog articles?
> 
> For example, headings should be sentence case, and new concepts should be italicized.

Thanks @sftim - I made some updates so this blog post will better adhere to the style guides.

jsturtevant commented on content/en/blog/_posts/Windows-Memory-Pressure-Eviction.md:
```diff
@@ -0,0 +1,85 @@
+
+Additionally, exposing the _Commit_ memory usage for individial Pods and Containers through the `/stats/summary` API is being worked on as part of the [cAdvisor-les, CRI-full Container and Pod Stats](https://github.com/kubernetes/enhancements/blob/master/keps/sig-node/2371-cri-pod-container-stats/README.md) enhancement.
+
+[windows-exporter](https://github.com/prometheus-community/windows_exporter) can also be configured to monitor and report on _Commit_ memory usage for Pods and containers today.
```
```suggestion
[windows-exporter](https://github.com/prometheus-community/windows_exporter) can also be configured to monitor and report on _Commit_ memory usage for Pods and Containers today.
```

It is upper cased in sentence above

sftim replied:
Only write Container in title case if you are referring to API level details and the lower case form would be confusing.

marosset replied:
I made the above mentioned Containers lowercase too

jsturtevant commented on content/en/blog/_posts/Windows-Memory-Pressure-Eviction.md:
```diff
@@ -0,0 +1,85 @@
+
+## Eviction thresholds
+
+By default, Windows nodes have a default "hard" eviction threshold of `--eviction-hard=memory.available<500Mi`. This means that if the system's availalbe memory is less than 500 Mi the Kubelet will try and start evicting pods.
```
This still missing bracket and availalbe spelling.

marosset replied:
See https://github.com/kubernetes/website/pull/48535#discussion_r1819508392 - we don't want the bracket there

jsturtevant commented:
/lgtm
for sig-windows, pending the small format updates

sftim commented on content/en/blog/_posts/Windows-Memory-Pressure-Eviction.md:
```diff
@@ -0,0 +1,85 @@
+
+Understanding the distinction between _Working Set_ and _Commit_ memory and being able to monitor both is critical for maintaining the stability of individual Pod workloads and the entire Windows node.
+
+Currently, `kubectl top` continues to report Working Set memory. As a result, a Pod may be evicted for exceeding its memory requests or encounter memory allocation failures due to exceeding its `CommitLimit`, even when the _WorkingSet_ memory usage appears to be well within the expected bounds. This discrepancy can lead to confusion.
+
+Starting with v1.31, Windows nodes now report the global `CommitTotal` and `CommitLimit` for each node through the `/stats/summary` API endpoint. This information is available under the `windows-global-commit-memory` syscontainer API resource.
```
This looks out of date.

marosset replied:
Which part specifically?
I don't think this is out of date.

sftim replied:
See https://github.com/kubernetes/website/pull/48535#discussion_r1872927644

tengqm commented on content/en/blog/_posts/Windows-Memory-Pressure-Eviction.md:
```diff
@@ -0,0 +1,85 @@
+
+_Commit_ memory, on the other hand, represents the total amount of virtual memory that has been reserved by processes - including memory that has been swapped out to disk. When memory is "committed", it means that the system has allocated the backing storage (either physical RAM or space in the paging file(s)) to ensure that the memory will be available if needed.
+
+Using global commit levels to measure available memory on Windows was implemented for two key reasons:
```
```suggestion
There are two reasons to using the global commit levels for measuring the available memory
on a Windows node:
```

tengqm commented on content/en/blog/_posts/Windows-Memory-Pressure-Eviction.md:
```diff
@@ -0,0 +1,85 @@
+
+Using global commit levels to measure available memory on Windows was implemented for two key reasons:
+
+1. **Memory limits for containers in Pods running on Windows nodes are enforced using Commit** (not WorkingSet) **memory**. Windows memory management enforces limits on processes based on the total amount of memory committed, not just the memory in use. 
```
```suggestion
1. **Memory limits for containers in Pods running on Windows nodes are enforced using Commit (not WorkingSet) memory**.
   Windows memory management enforces limits on processes based on the total amount of memory committed,
   not just the memory in use. 
```

tengqm commented on content/en/blog/_posts/Windows-Memory-Pressure-Eviction.md:
```diff
@@ -0,0 +1,85 @@
+Using global commit levels to measure available memory on Windows was implemented for two key reasons:
+
+1. **Memory limits for containers in Pods running on Windows nodes are enforced using Commit** (not WorkingSet) **memory**. Windows memory management enforces limits on processes based on the total amount of memory committed, not just the memory in use. 
+This also means that memory limits specified for containers in Kubernetes Pods are enforced on the total amount of virtual memory committed by all processes in the container. Consistency in how memory usage is tracked and enforced is essential for workload stability.
```
```suggestion
   This also means that memory limits specified for containers in Kubernetes Pods are enforced
   on the total amount of virtual memory committed by all processes in the container. 
   Consistency in how memory usage is tracked and enforced is essential for workload stability.
```

tengqm commented on content/en/blog/_posts/Windows-Memory-Pressure-Eviction.md:
```diff
@@ -0,0 +1,85 @@
+
+## Eviction thresholds
+
+By default, Windows nodes have a "hard" eviction threshold of 500Mi (kubelet setting `--eviction-hard=memory.available<500Mi`). This means that if the system's availble commit memory is less than 500 Mi the Kubelet will try and start evicting pods.
```
```suggestion
By default, Windows nodes have a "hard" eviction threshold of 500Mi
(kubelet setting `--eviction-hard=memory.available<500Mi`).
This means that if the system's availble commit memory is less than 500 Mi,
the Kubelet will try to evict pods.
```

tengqm commented on content/en/blog/_posts/Windows-Memory-Pressure-Eviction.md:
```diff
@@ -0,0 +1,85 @@
+
+By default, Windows nodes have a "hard" eviction threshold of 500Mi (kubelet setting `--eviction-hard=memory.available<500Mi`). This means that if the system's availble commit memory is less than 500 Mi the Kubelet will try and start evicting pods.
+
+For more information on configuring memory-pressure based eviction and understanding the eviction signals and thresholds, refer to the official Kubernetes documentation on [memory-pressure-eviction](https://kubernetes.io/docs/concepts/scheduling-eviction/node-pressure-eviction/#eviction-signals-and-thresholds).
```
```suggestion
For more information on configuring memory-pressure based eviction and understanding the eviction signals and thresholds,
refer to the official Kubernetes documentation on [memory-pressure-eviction](/docs/concepts/scheduling-eviction/node-pressure-eviction/#eviction-signals-and-thresholds).
```

tengqm commented on content/en/blog/_posts/Windows-Memory-Pressure-Eviction.md:
```diff
@@ -0,0 +1,85 @@
+
+## Monitoring commit memory for Windows nodes
+
+Understanding the distinction between _Working Set_ and _Commit_ memory and being able to monitor both is critical for maintaining the stability of individual Pod workloads and the entire Windows node.
```
"Working Set" or "WorkingSet"?
Please be consistent in the article.

marosset replied:
Sorry, i thought I updated everything. I'll scan everything again.

tengqm commented on content/en/blog/_posts/Windows-Memory-Pressure-Eviction.md:
```diff
@@ -0,0 +1,85 @@
+
+Currently, `kubectl top` continues to report Working Set memory. As a result, a Pod may be evicted for exceeding its memory requests or encounter memory allocation failures due to exceeding its `CommitLimit`, even when the _WorkingSet_ memory usage appears to be well within the expected bounds. This discrepancy can lead to confusion.
+
+Starting with v1.31, Windows nodes now report the global `CommitTotal` and `CommitLimit` for each node through the `/stats/summary` API endpoint. This information is available under the `windows-global-commit-memory` syscontainer API resource.
```
Maybe you want to remove "now".

The "syscontainer API resource" is a little bit confusing to me.

marosset replied:
I reworked it to 
> This information is available under the `windows-global-commit-memory` named object in the `systemContainers` collection:

Does that make more sense?

tengqm commented on content/en/blog/_posts/Windows-Memory-Pressure-Eviction.md:
````diff
@@ -0,0 +1,85 @@
+
+```
+
+Additionally, exposing the _Commit_ memory usage for individial Pods and containers through the `/stats/summary` API is being worked on as part of the [cAdvisor-les, CRI-full Container and Pod Stats](https://github.com/kubernetes/enhancements/blob/master/keps/sig-node/2371-cri-pod-container-stats/README.md) enhancement.
````
cAdvisor-less ?

marosset replied:
fixed

tengqm commented:
/lgtm
/label tide/merge-method-squash

marosset commented:
/hold 
I'm not sure what the process for assigning a release date is for blogs but this one still has TBD

sftim reviewed (commented):
Thanks @marosset 
- Please set the publication date to Tuesday 11th December 2024 (we'll update that after it merges). This is a change to the article path in source and to the front matter.
- Please add `draft: true` to the page front matter

(nit)
Ideally, also avoid writing component names such as kubelet in anything other than all-lowercase.

sftim commented:
/lgtm cancel

OK to LGTM this if it has a future publication date agreed with release comms, or it it's marked as draft and the publication date is set to the planned release date for K8s 1.32

Otherwise, reviewers should hold off there.

marosset commented:
@sftim - I made the following updates

- renamed 1 instance of Kubelet to kubelet,
- updated the arcticle name in source
- set `draft: true` in the front matter
- set publication date to `2024-12-11` in the front matter.
Let me know if there is anything else needed!

marosset commented:
@sftim / @tengqm - Is there any other action required from me for this?
Thanks!

sftim commented:
I don't see this on https://github.com/orgs/kubernetes/projects/195/views/4

The advice in https://github.com/kubernetes/website/pull/48535#issuecomment-2468728812 was conditional on this being accepted for the post-release comms series. @marosset can you liaise there (`#release-comms` on Slack)? It may be that this becomes an informal article that goes out after the formal post release comms.

marosset commented:
> I don't see this on https://github.com/orgs/kubernetes/projects/195/views/4
> 
> The advice in [#48535 (comment)](https://github.com/kubernetes/website/pull/48535#issuecomment-2468728812) was conditional on this being accepted for the post-release comms series. @marosset can you liaise there (`#release-comms` on Slack)? It may be that this becomes an informal article that goes out after the formal post release comms.

I think it would make more sense for this to go out as an informal article.
This is going more in-depth into work that was done in v1.31 and prior, not 1.32 / the current release.
Is there anything that needs to happen if we go that route?

sftim commented on content/en/blog/_posts/2024-12-11-windows-memory-pressure-eviction.md:
```diff
@@ -0,0 +1,95 @@
+---
+layout: blog
+title: "Windows Memory Pressure Eviction"
+slug: windows-memory-pressure-eviction
+date: 2024-12-11
+draft: true
+author: >
+  Mark Rossetti (Microsoft)
```
OK, try this:
```suggestion
layout: blog
title: "Windows Memory Pressure Eviction"
slug: windows-memory-pressure-eviction
date: 2024-12-09
author: >
  Mark Rossetti (Microsoft)
```
and change the path in the front matter to match the new date.

sftim commented on content/en/blog/_posts/2024-12-11-windows-memory-pressure-eviction.md:
````diff
@@ -0,0 +1,95 @@
+
+Starting with v1.31, Windows nodes report the global `CommitTotal` and `CommitLimit` for each node through the `/stats/summary` API endpoint. This information is available under the `windows-global-commit-memory` named object in the `systemContainers` collection:
+
+```json=
````
Use `console` here because this is not a JSON document

sftim commented on content/en/blog/_posts/2024-12-11-windows-memory-pressure-eviction.md:
```diff
@@ -0,0 +1,95 @@
+
+Understanding the distinction between _Working Set_ and _Commit_ memory and being able to monitor both is critical for maintaining the stability of individual Pod workloads and the entire Windows node.
+
+Currently, `kubectl top` continues to report Working Set memory. As a result, a Pod may be evicted for exceeding its memory requests or encounter memory allocation failures due to exceeding its `CommitLimit`, even when the _Working Set_ memory usage appears to be well within the expected bounds. This discrepancy can lead to confusion.
```
Did you mean:
```suggestion
Before v1.31, the values you saw from `kubectl top` reported Working Set memory. As a result, a Pod could have been evicted for exceeding its memory requests, or a Pod might have encountered memory allocation failures (due to exceeding its `CommitLimit`) - and this could happen even when the _Working Set_ memory usage appeared to be well within the expected bounds. This discrepancy could have led to confusion.
```

It's not clear if v1.31 improved things.

sftim commented:
To be clear: this won't be a post-release blog article. I'd got mixed up and thought this was part of formal post-release comms.

sftim commented:
/hold

It'll need a new release date and we'll need to update this to avoid people inferring that the feature was part of the v1.32 release.

Sorry about the wait on reviews here.

OK to unhold once a publication date is agreed.

sftim commented:
Actually, this is OK to merge as a draft and then a small follow-up PR can set a (future) publication date, removing the `draft: true` from the front matter at the same time.

/hold cancel

@marosset would you have capacity to look at the pending feedback? There are only a couple of points to cover.
